File: scripts/snapshot.py
"""Deterministic filesystem fingerprinting for backup change detection."""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from pathlib import Path


@dataclass
class Snapshot:
    fingerprint: str
    file_count: int
    total_bytes: int


def _add_entries(
    abs_path: Path,
    rel_path: str,
    entries: list[str],
    counters: dict[str, int],
) -> None:
    if abs_path.is_symlink():
        mode = oct(abs_path.lstat().st_mode & 0o777).replace("0o", "").zfill(3)
        counters["file_count"] += 1
        entries.append(f"symlink {rel_path} mode={mode} -> {abs_path.readlink()}")
        return

    if abs_path.is_dir():
        mode = oct(abs_path.stat().st_mode & 0o777).replace("0o", "").zfill(3)
        entries.append(f"dir {rel_path} mode={mode}")
        for child in sorted(abs_path.iterdir(), key=lambda c: c.name):
            child_rel = f"{rel_path}/{child.name}" if rel_path else child.name
            _add_entries(child, child_rel, entries, counters)
        return

    if abs_path.is_file():
        stat = abs_path.stat()
        mode = oct(stat.st_mode & 0o777).replace("0o", "").zfill(3)
        counters["file_count"] += 1
        counters["total_bytes"] += stat.st_size
        file_hash = hashlib.sha256(abs_path.read_bytes()).hexdigest()
        entries.append(
            f"file {rel_path} mode={mode} size={stat.st_size} sha256={file_hash}"
        )
        return

    mode = oct(abs_path.stat().st_mode & 0o777).replace("0o", "").zfill(3)
    entries.append(f"other {rel_path} mode={mode}")


def source_snapshot(root: str | Path, relative_paths: list[str]) -> Snapshot:
    """Compute a deterministic fingerprint over a set of paths relative to a root."""
    root_path = Path(root)
    entries: list[str] = []
    counters = {"file_count": 0, "total_bytes": 0}

    for rel_path in relative_paths:
        _add_entries(root_path / rel_path, rel_path, entries, counters)

    fingerprint = hashlib.sha256("\n".join(entries).encode()).hexdigest()
    return Snapshot(
        fingerprint=fingerprint,
        file_count=counters["file_count"],
        total_bytes=counters["total_bytes"],
    )


def directory_snapshot(root: str | Path) -> Snapshot:
    """Compute a fingerprint for all contents of a single directory root."""
    root_path = Path(root)
    entries: list[str] = []
    counters = {"file_count": 0, "total_bytes": 0}

    for child in sorted(root_path.iterdir(), key=lambda c: c.name):
        _add_entries(child, child.name, entries, counters)

    fingerprint = hashlib.sha256("\n".join(entries).encode()).hexdigest()
    return Snapshot(
        fingerprint=fingerprint,
        file_count=counters["file_count"],
        total_bytes=counters["total_bytes"],
    )

File: packages/core/src/snapshot.ts
import { createHash } from 'node:crypto'
import fs from 'node:fs'
import path from 'node:path'

import { type Snapshot, SnapshotSchema } from './schema'

const octalMode = (mode: number) => (mode & 0o777).toString(8).padStart(3, '0')

const sha256Hex = (value: Buffer | string) => createHash('sha256').update(value).digest('hex')

const addEntries = (
  absolutePath: string,
  relativePath: string,
  entries: Array<string>,
  counters: { fileCount: number; totalBytes: number },
) => {
  const lstat = fs.lstatSync(absolutePath)

  if (lstat.isSymbolicLink()) {
    counters.fileCount += 1
    entries.push(
      `symlink ${relativePath} mode=${octalMode(lstat.mode)} -> ${fs.readlinkSync(absolutePath)}`,
    )
    return
  }

  if (lstat.isDirectory()) {
    entries.push(`dir ${relativePath} mode=${octalMode(lstat.mode)}`)
    for (const child of fs.readdirSync(absolutePath).sort()) {
      addEntries(
        path.join(absolutePath, child),
        relativePath ? `${relativePath}/${child}` : child,
        entries,
        counters,
      )
    }
    return
  }

  if (lstat.isFile()) {
    const bytes = fs.readFileSync(absolutePath)
    counters.fileCount += 1
    counters.totalBytes += bytes.length
    entries.push(
      `file ${relativePath} mode=${octalMode(lstat.mode)} size=${bytes.length} sha256=${sha256Hex(bytes)}`,
    )
    return
  }

  entries.push(`other ${relativePath} mode=${octalMode(lstat.mode)}`)
}

export const sourceSnapshot = (root: string, relativePaths: ReadonlyArray<string>): Snapshot => {
  const entries: Array<string> = []
  const counters = { fileCount: 0, totalBytes: 0 }

  for (const relativePath of relativePaths) {
    addEntries(path.join(root, relativePath), relativePath, entries, counters)
  }

  return SnapshotSchema.parse({
    fileCount: counters.fileCount,
    fingerprint: sha256Hex(entries.join('\n')),
    totalBytes: counters.totalBytes,
  })
}

export const directorySnapshot = (root: string): Snapshot => {
  const entries: Array<string> = []
  const counters = { fileCount: 0, totalBytes: 0 }

  for (const child of fs.readdirSync(root).sort()) {
    addEntries(path.join(root, child), child, entries, counters)
  }

  return SnapshotSchema.parse({
    fileCount: counters.fileCount,
    fingerprint: sha256Hex(entries.join('\n')),
    totalBytes: counters.totalBytes,
  })
}
